helix-core/src/indent.rs

```diff
@@ -237,38 +237,58 @@ fn get_first_in_line(mut node: Node, new_line_byte_pos: Option<usize>) -> Vec<bo
 /// This is usually constructed in one of 2 ways:
 /// - Successively add indent captures to get the (added) indent from a single line
 /// - Successively add the indent results for each line
-#[derive(Default)]
+#[derive(Default, Debug, PartialEq, Eq, Clone)]
 pub struct Indentation {
     /// The total indent (the number of indent levels) is defined as max(0, indent-outdent).
     /// The string that this results in depends on the indent style (spaces or tabs, etc.)
     indent: usize,
+    indent_always: usize,
     outdent: usize,
+    outdent_always: usize,
 }
 impl Indentation {
     /// Add some other [Indentation] to this.
     /// The added indent should be the total added indent from one line
     fn add_line(&mut self, added: &Indentation) {
-        if added.indent > 0 && added.outdent == 0 {
-            self.indent += 1;
-        } else if added.outdent > 0 && added.indent == 0 {
-            self.outdent += 1;
-        }
+        self.indent += added.indent;
+        self.indent_always += added.indent_always;
+        self.outdent += added.outdent;
+        self.outdent_always += added.outdent_always;
     }
+
     /// Add an indent capture to this indent.
     /// All the captures that are added in this way should be on the same line.
     fn add_capture(&mut self, added: IndentCaptureType) {
         match added {
             IndentCaptureType::Indent => {
-                self.indent = 1;
+                if self.indent_always == 0 {
+                    self.indent = 1;
+                }
+            }
+            IndentCaptureType::IndentAlways => {
+                // any time we encounter an `indent.always` on the same line, we
+                // want to cancel out all regular indents
+                self.indent_always += 1;
+                self.indent = 0;
```

Comment:
Yeah, it's because of when we're rendering the indent here:

https://github.com/helix-editor/helix/pull/6768/files#diff-44db3a32fdd620ed51f6953534c6ed5df3d2966010cc86aaf13448e9bbb65f07R287-R288

The issue is that during the indent capture search, we accumulate the indents across two dimensions:

* `add_capture` accumulates *within* lines; we want to ensure that all `indent.always` captures are always additive. `indent` captures within lines are more like a boolean—any indent capture encountered, no matter which type, means any further `indent` captures have no effect
* `add_line` accumulates *across* lines. The purpose is to combine indent levels of lines higher up in the syntax tree, so indent levels are always additive

In order to ensure both of these are true, `indent` captures are still stored as a `usize`, despite their behavior when accumulating within lines, because we need to be able to sum them across lines. But if we don't reset the `indent` field here, then we can't tell the difference between `indent` captures within and across lines when we render the final indent level.

The API is a bit confusing and hard to follow because it uses the same type and object instance to compute indents within and across lines, and the only thing ensuring that everything works is that it so happens that we only use the `add_line` and `add_capture` methods where we're supposed to.

But fixing that would need a larger refactor and I wasn't quite up for that.

Reply:
Yeah, the code would definitely be easier to read if we had different types for the total indent and the indent for one line.

> But if we don't reset the `indent` field here, then we can't tell the difference between `indent` captures within and across lines when we render the final indent level.

Is this really true? Even if we don't reset it, the `add_capture` never increases the `indent` field beyond 1, so multiple `@indent` captures on the same line should still cancel out. The behavior should only be different when `@indent.always` and `@indent` appear on the same line. They could either stack (i.e. produce a total indent of 2) or the `@indent` could be ignored, producing a total indent of 1 (the current behavior).

From the queries, it's not obvious to me whether this can actually happen (since the only `@indent` capture is for the `block_scalar` node). If it doesn't, I think the first behavior makes more sense (but of course this is somewhat subjective).

Reply:
Yeah, this was part of the discussion starting [here](https://github.com/helix-editor/helix/pull/6768#issuecomment-1595693901) that centered around what should happen when we encounter both on the same line; the problem was it could stack in one case and not the other, depending on the order they were encountered.

For consistency, we should choose one or the other. But which one is I think the issue at hand. The way I see it, `@indent.always` is just "another kind of `@indent`", so if we should combine multiple `@indent`s on the same line, then we should include `@indent.always` in that calculation.

But I can see how the other way would make sense too.

Reply:
> For consistency, we should choose one or the other. But which one is I think the issue at hand.

I definitely agree.



> But which one is I think the issue at hand. The way I see it, `@indent.always` is just "another kind of `@indent`", so if we should combine multiple `@indent`s on the same line, then we should include `@indent.always` in that calculation.

Yeah that makes sense. I think that to me `@indent.always` seemed like something separate from a regular `@indent`, but (especially considering the naming) your intuition is probably better. Thanks for explaining it to me.